### DISCUSSION TITLE
bool类型时是否has_config返回true更合理？

### DISCUSSION THREAD

**Discussion by PeakRacing** [xmake-io/xmake#4873] [Q&A]
has_config更像是该选项是否配置生效的意思？如果是这样那bool类型配置成false是否has_config返回true更合理，然后通过get_config获取具体值，和字符串处理一样，这样理解上来看更合理，使用也更统一

**Comment by waruqi** [comment]
目前就是返回 true https://github.com/xmake-io/xmake/blob/e1d5f432ab3d0fac43ee6a4852efe7b99f6ef428/xmake/core/project/config.lua#L300

至于 false，目前返回的是 nil，发不返回 false 其实没啥，走 `if has_config()` 结果都一样，当然要返回 false 也可以

**Reply by PeakRacing** [reply]
nil/false都一样，我想问的其实是只有字符串配置才返回true，bool类型配置成false时候has_config不会返回true，这样就区分不开是config还未生效还是就配置成false了，所以就算bool类型配置成false，has_config也返回true去告诉配置已经生效，然后在通过get_config获取具体值，这样还可以和字符串使用统一，也便于理解，也消除了是否配置生效的疑问

**Reply by waruqi** [reply]
不，bool 类型，配置成 false ，就是没生效。。等价于 `xmake f --xxx=n` ，不要把它当成生效使用，非要判断是否为 false ，就自己走 get_config 自己判断

**Reply by PeakRacing** [reply]
走get_config判断无法区分是否配置生效，都是nil/faslse，然后不得已都走字符串配置， 而且总觉得这样和has_config名有点歧义了，哈哈

**Reply by waruqi** [reply]
搞不懂你要搞啥，走 bool 类型，true 生效，false/nil 不生效。。不就完事了么。。还想搞啥

**Reply by PeakRacing** [reply]
我要根据true/false配置而且不能都配置一遍，是要明确是true还是false，刚编译阶段配置是未生效的，就算是true也一样得到的是nil/false，我要是根据has_config/get_config那直接配置false了

**Reply by waruqi** [reply]
> 刚编译阶段配置是未生效的，就算是true也一样得到的是nil/false

那是因为你没配置，`set_default(true)` ，默认当然是 nil

bool option 没有所谓的生效不生效，只有 true 启用，nil/false 不启用。。默认不设置 set_default ，就是 nil，会走 option 自动检测，检测不到，就是禁用

反正 bool option 只有两种状态，跟是否明确 true/false 关系。。原本就只有 true or false 两种状态

**Reply by PeakRacing** [reply]
未生效时候打印看得出来就是nil，生效了才是true/false，字符串也一样，未生效时候是nil，生效之后才是默认设置的字符串，做个打印直接就看出来了
![image](https://github.com/xmake-io/xmake/assets/24759100/dc2800db-bb95-4ed9-9ca1-275a8e55b47f)
![image](https://github.com/xmake-io/xmake/assets/24759100/104ddf88-d37e-42dd-8ff7-28805df26e2f)
就算是字符串也要等生效才能使用
bool型更是
![image](https://github.com/xmake-io/xmake/assets/24759100/8a541de4-f244-4812-ab43-00edb109f048)
![image](https://github.com/xmake-io/xmake/assets/24759100/df97e4ee-c37c-4110-880d-8e36a45e4acc)

**Reply by waruqi** [reply]
> 未生效时候打印看得出来就是nil，生效了才是true/false，字符串也一样，未生效时候是nil，生效之后才是默认设置的字符串，做个打印直接就看出来了

不是，描述域不要 print 任何东西，也不要管 has_config 返回的是 nil 还是 false。。只管 `if has_config("xx")` 就行了，进 if 就是启动，不进 if 就是禁用，不要去纠结返回值。。无意义。。也不要依赖它返回的是 nil 还是 false 。。对于 xmake ，返回的 nil 跟 false 是相同状态，就是禁用。

**Reply by PeakRacing** [reply]
我打印只是看下状态，这里打印和执行是一样的并不会影响，has_config("xx")就是有延迟的，不管是否设置上来一定是nil，而且bool如果设置false那has_config("xx")就一直是nil，根本判断不出来是没配置还是配置的false
就比如
```lua
option("test", {default = false, showmenu = true,description = "Enable or disable low speed mode , more memory"})
add_options("test")

if has_config("test") then
    if get_config("test") == true then
        --配置xxx
    elseif get_config("test") == false then
        --配置xxx
    end
end 
```
我如果配置true的情况下，可以进到配置，如果我设置默认false，那就永远是nil，配置不到 get_config("test") == false 判断

**Reply by waruqi** [reply]
很正常，你用法不对，如果是 bool config 。。你只能用 `if has_config("test") then` 。。即使你用 `get_config("xxx")`，

也只能 

```lua
if get_config("test") then
else
end
```

or 

```lua
if not get_config("test") then
else
end
```

不要去判断 false，只有两种状态，启用或者禁用。

所以对于 bool option ，请仅用 has_config 。。get_config 那个主要是给 string option 用的

**Reply by PeakRacing** [reply]
不对，我上面说了，我要根据true/false来实现不同配置，使用has_config 那false一定会被配置了，不管我是否配置了bool值，都会去配置一遍false

**Reply by waruqi** [reply]
> 使用has_config 那false一定会被配置了，不管我是否配置了bool值，都会去配置一遍false

不懂你说的啥，你没配置，默认就是 nil ，也就是 false 。。也就是 option 被禁用状态。。但你不能直接去判断 false 值，得用 has_config 或者 not has_config 判断是否被禁用

**Reply by PeakRacing** [reply]
我没配置或者配置成false，has_config 会一直只返回nil，我只有配置了true，has_config 会先返回nil后返回true，无法正确根据返回配置自己的相关配置

**Reply by waruqi** [reply]
> 我没配置或者配置成false，has_config 会一直只返回nil

这不是正常的么，刚说了 bool option 只有启用和禁用两种状态，没配置 nil 你就把它当做 false 。。不要直接判断 false，用 `if has_config` 就行了

> has_config 会先返回nil后返回true，无法正确根据返回配置自己的相关配置

这就是我不让你 描述域用 print 的原因。。文档里也有明确说明。。不要用 print dump 任何描述域状态，没任何意义。。

你不要管 先返回 nil 后返回 true 没任何意义。。。尽管用 `if has_config` 就行了。。

描述域配置，会被重复解析多次，不同阶段的解析，print 的状态都是不同的，首次解析 option 都还没被解析，肯定都是 nil，仅仅只是解析 xmake --help 的早期菜单项，但这不影响后期 `if has_config` 实际配置 和 target 的配置结果。。

中途还有各种解析，都有不同的意义。。你不管怎么 print 。。看到什么 config 值 ，都是没有任何意义的，除非你完全了解 xmake 内部机制。。实在理解不了，那我之后只能把描述域的 print 给禁用了。

<img width="865" alt="image" src="https://github.com/xmake-io/xmake/assets/151335/4cc6634a-e550-4b9f-8692-ead092508f75">

**Reply by waruqi** [reply]
你不要管中间状态，has_config/get_config 返回了啥。。反正就这么用

```lua
if has_config("test") then
    -- enabled
else
   -- disabled
end
```

没配置，就是 disabled，配置 `xmake f --test=n` 也是 disabled，配置 `xmake f --test=y` 就是 enabled。。

这要这个对上，就是没问题的。。如果这个不对。。就提 issues bug 过来，并提供复现过程和详细配置

**Reply by waruqi** [reply]
另外，没配置，还有一种情况就是会走自动检测，如果你配置了 add_cfuncs add_cincludes 这些。.检测通过就是 enabled，检测不通过 默认也是 disabled

**Reply by PeakRacing** [reply]
这么看是和我理解的执行不一样，有一些配置域好像并没有执行我想的配置，后面我学习下，先做下测试
